Refactor presences script

fgrehm opened the pull request:
This builds on top of a pairing session I had with @cuducos and it is made of:

- The extraction of some utility methods to be shared with other scripts
- Lots of cleanup
- More information extracted / normalized for the Deputies dataset:
  - `registration` was renamed to `congressperson_document`
  - `name` was renamed to `congressperson_name`
  - New fields: `picture_url`, `gender`, `phone_number`, `email`
- More information extracted / normalized for the Presences dataset:
  - `registration` was renamed to `congressperson_document`
  - `presence` was translated to english
  - New fields: `term`, `present_on_day`, `justification`
- Logic for fetching presences made more robust by issuing retries and dealing with the API "flakyness"
- On the SessionStartTimes, the `session` field was renamed to `description`

fgrehm commented:
Almost forgot, here's da data 😋 

- [2017-01-16-deputies.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/711118/2017-01-16-deputies.xz.zip)
- [2017-01-16-presences.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/711119/2017-01-16-presences.xz.zip)
- [2017-01-16-session-start-times.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/711120/2017-01-16-session-start-times.xz.zip)

And the [execution logs](https://gist.github.com/fgrehm/6e1c5f6939eba0847a77ada3c07b9f07)

cuducos commented on serenata_toolbox/presences.py:
```diff
@@ -1,143 +1,165 @@
+        df = pd.DataFrame(records, columns=(
+            'term',
+            'congressperson_document',
+            # TODO: Rename to something else to avoid confusion (maybe just congressperson?)
```
Not sure what you mean here — the filed holds the congressperson's name, doesn't it? Or is it the string with party and state? If so, we can try to extract with a regex or stuff…

fgrehm replied:
Yeah, `congressperson_name` looks like `Luana Costa-PSB/MA`, but we already have separate columns for state and party so I guess the best we can do is either rename it to something else to avoid confusion while keeping the dataset as close to the raw source as possible; or we can remove the party and state from the field and keep just the name around.

@Irio @cuducos Thoughts?

cuducos replied:
I'd vote for _remove the party and state_, but @irio is smarter than me when it come to data treatment and cleaning…

Irio replied:
I like the idea of having field names close to original sources and documentation for addressing issues caused by it.

cuducos commented on serenata_toolbox/presences.py:
```diff
@@ -1,143 +1,165 @@
+                return urllib.request.urlopen(url, data=None, timeout=10)
+            except urllib.error.HTTPError as err:
+                print("HTTP Error", err.code, "when loading URL", url)
+                # 500 seems to be the error code for "no data found for the
```
Ow boy… I don'r remember where, but another service from the government used to respond `HTTP 200: Arquivo não encontrado` hahaha…

cuducos commented on serenata_toolbox/presences.py:
```diff
@@ -1,143 +1,165 @@
+                    print("Trying again", attempts)
+                else:
+                    print("FAIL")
+            except socket.error as socketerror:
```
Nice!

cuducos commented on serenata_toolbox/session_start_times.py:
```diff
@@ -0,0 +1,57 @@
+    def fetch(self, pivot, session_dates):
+        """
+        :param pivot: (int) a congressperson document to use as a pivot for scraping the data
+        :param session_dates: (str) a list of datetime objects to fetch the start times for
```
> :param session_dates: (str) a list of datetime objects…

Is it a `str`, a `list` or `datetime`?

fgrehm replied:
Its an array / list of datetimes, how do we document that? 🤔

cuducos replied:
I'd use use `:param session_dates: (list) datetime objects to fetch the start times for…`

cuducos commented on serenata_toolbox/deputies.py:
```diff
@@ -0,0 +1,82 @@
+import pandas as pd
+
+from serenata_toolbox import datasets
+from serenata_toolbox.cleanup import xml_extract_text, \
```
Totally 💅 but… usually in Python styles goes for parenthesis in that case:

```python
from serenata_toolbox.cleanup import (
    xml_extract_date,
    xml_extract_datetime,
    xml_extract_text
)
```

fgrehm commented:
@cuducos @Irio @jtemporal  I've finally had a chance to get back to this 😅 

Please take a look at the latest changes and LMK what u think about them. I've also fetched a new version of the datasets that would be nice to be uploaded to S3:

- [2017-04-19-session-start-times.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/935430/2017-04-19-session-start-times.xz.zip)
- [2017-04-19-presences.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/935432/2017-04-19-presences.xz.zip)
- [2017-04-19-deputies.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/935431/2017-04-19-deputies.xz.zip)
- [Execution logs](https://gist.github.com/fgrehm/bfe768a668fd5e93bff04f7b2c97ed9f)

_(as usual, github does not let me upload `.xz` files so I had to `zip` them in order to upload, make sure to unzip and upload the `xz` only_ 😄 _)_

cuducos commented on serenata_toolbox/cleanup.py:
```diff
@@ -0,0 +1,42 @@
+        text = text.strip()
+    return text
+
+def xml_extract_date(node, xpath):
```
What about that:

```python
def xml_extract_date(node, xpath, date_format='%d/%m/%Y'):
    …
```
Then this function can be used in a more generic way if needed ; )

cuducos commented on serenata_toolbox/cleanup.py:
```diff
@@ -0,0 +1,42 @@
+    """
+    return datetime.strptime(xml_extract_text(node, xpath), "%d/%m/%Y")
+
+def xml_extract_datetime(node, xpath):
```
What about that:

```python
def xml_extract_datetime(node, xpath, date_format='%d/%m/%Y %H:%M:%S'):
    …
```
Then this function can be used in a more generic way if needed ; )

cuducos commented:
Just in case I just uploaded these datasets to S3:

* `2017-04-19-session-start-times.xz.zip`
* `2017-04-19-presences.xz.zip`
* `2017-04-19-deputies.xz.zip`
* `2017-04-21-speeches.xz.zip`

fgrehm commented:
Here's another round of updates based on the new code:
- [2017-05-29-speeches.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/1036972/2017-05-29-speeches.xz.zip)
- [2017-05-29-session-start-times.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/1036974/2017-05-29-session-start-times.xz.zip)
- [2017-05-29-presences.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/1036971/2017-05-29-presences.xz.zip)
- [2017-05-29-deputies.xz.zip](https://github.com/datasciencebr/serenata-toolbox/files/1036973/2017-05-29-deputies.xz.zip)
- Logs at https://gist.github.com/fgrehm/d12219588f2ea4eb648c9aa036ee6806

fgrehm commented:
@cuducos I added some integration tests for the new datasets and some unit tests for the helper methods introduced. Do u think we need some additional testing for the datasets or the ~~sanity checks~~ integration tests I introduced are enough?

cuducos commented on serenata_toolbox/chamber_of_deputies/presences_dataset.py:
```diff
@@ -0,0 +1,173 @@
+
+            # Let's not slow down our integration tests
+            if os.environ.get('RUN_INTEGRATION_TESTS') != '1':
+                time.sleep(2)
```
I think this is not a good practice… let's try to code that into the test suite, not in the script itself? I can help… maybe we pair, maybe I can try to code it in another branch… what do you think?

cuducos replied:
This sounds like a bit of code smell… it's not clear why we need this only in integration tests…  if you clarify why we need that I can help you find a better testing strategy ; )

fgrehm replied:
Yeah, theses `sleep`s have been around for a while 😊 The idea is to prevent us from hammering camara's API with requests. The reason behind having them baked into the code is that the first few times I ran these scripts their server would eventually start 💥 ing like half way through scraping.

I guess I could just pass it in as a parameter to the dataset constructor 🤔 Will give that a try

cuducos replied:
Yeah… I mean… the way it's coded it looks like it's a test dependency, not a thing to be used in production… following your description, a variable passed in the instantiation sounds really good ; )

cuducos commented on serenata_toolbox/datasets/downloader.py:
```diff
@@ -1,5 +1,6 @@
 import asyncio
 import os
+from datetime import datetime
```
Why do we need it here?

fgrehm replied:
Probably bad merge, will revert

cuducos commented:
As landscape.io is temporally not running in CI mode, I ran `prospector` locally, this is the report if it helps:

```
Messages
========

chamber_of_deputies/deputies_dataset.py
  Line: 12
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 71
    pep8: E302 / expected 2 blank lines, found 1 (col 1)

chamber_of_deputies/presences_dataset.py
  Line: 9
    pyflakes: F401 / 'serenata_toolbox.datasets.helpers.xml_extract_date' imported but unused (col 1)
  Line: 17
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 20
    pep8: E501 / line too long (94 > 79 characters) (col 80)
  Line: 33
    pep8: E501 / line too long (108 > 79 characters) (col 80)
  Line: 55
    pep8: E501 / line too long (84 > 79 characters) (col 80)
  Line: 56
    pep8: E501 / line too long (87 > 79 characters) (col 80)
  Line: 59
    pep8: E711 / comparison to None should be 'if cond is None:' (col 20)
  Line: 100
    pep8: E303 / too many blank lines (2) (col 5)
  Line: 170
    pep8: E501 / line too long (89 > 79 characters) (col 80)
  Line: 171
    pep8: E501 / line too long (89 > 79 characters) (col 80)

chamber_of_deputies/session_start_times_dataset.py
  Line: 7
    pyflakes: F401 / 'serenata_toolbox.datasets.helpers.translate_column' imported but unused (col 1)
  Line: 14
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 16
    pep8: E501 / line too long (86 > 79 characters) (col 80)
  Line: 24
    pep8: E501 / line too long (93 > 79 characters) (col 80)
  Line: 25
    pep8: E501 / line too long (82 > 79 characters) (col 80)
  Line: 39
    pep8: E501 / line too long (92 > 79 characters) (col 80)
  Line: 48
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 51
    pep8: E501 / line too long (87 > 79 characters) (col 80)
  Line: 59
    pep8: E501 / line too long (83 > 79 characters) (col 80)

chamber_of_deputies/speeches_dataset.py
  Line: 4
    pyflakes: F401 / 'datetime.datetime' imported but unused (col 1)
  Line: 15
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 18
    pep8: E501 / line too long (91 > 79 characters) (col 80)
  Line: 58
    pep8: E221 / multiple spaces before operator (col 24)
  Line: 63
    pep8: E221 / multiple spaces before operator (col 39)
    pep8: E501 / line too long (84 > 79 characters) (col 80)
  Line: 64
    pep8: E221 / multiple spaces before operator (col 40)
    pep8: E501 / line too long (82 > 79 characters) (col 80)
  Line: 65
    pep8: E501 / line too long (85 > 79 characters) (col 80)
  Line: 66
    pep8: E501 / line too long (80 > 79 characters) (col 80)
  Line: 69
    pep8: E501 / line too long (94 > 79 characters) (col 80)
  Line: 71
    pep8: E501 / line too long (105 > 79 characters) (col 80)
  Line: 72
    pep8: E126 / continuation line over-indented for hanging indent (col 33)
  Line: 79
    pep8: E221 / multiple spaces before operator (col 36)
    pep8: E501 / line too long (83 > 79 characters) (col 80)
  Line: 80
    pep8: E501 / line too long (85 > 79 characters) (col 80)
  Line: 97
    pep8: E302 / expected 2 blank lines, found 1 (col 1)

datasets/helpers.py
  Line: 13
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 19
    pep8: E711 / comparison to None should be 'if cond is not None:' (col 13)
  Line: 23
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 30
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 39
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
  Line: 43
    pep8: E501 / line too long (92 > 79 characters) (col 80)
  Line: 47
    pep8: E128 / continuation line under-indented for visual indent (col 20)
  Line: 50
    pep8: E302 / expected 2 blank lines, found 1 (col 1)
```

fgrehm commented:
@cuducos all done 😄

cuducos commented:
🎉 Yay! Many tks @fgrehm
